### ISSUE TITLE
Enums seem to have been broken on the Python-side since the support was added

### ISSUE THREAD

**Issue by alexer** [iovisor/bcc#2615]
Using structs that have enums in them generates JSON that `BPF._decode_table_type()` can't handle:
```
from bcc import BPF

b = BPF(text="""
struct S {
        enum {FOO} x;
};

BPF_HISTOGRAM(enumtest, struct S);
""")

typetest = b.get_table("enumtest")
```

The resulting JSON is `["S", [["x",["E", ["FOO"], "enum"]] ], "struct_packed"]` and the error is:
```
Traceback (most recent call last):
  File "bcc_enum.py", line 11, in <module>
    typetest = b.get_table("enumtest")
  File "/usr/lib64/python3.6/site-packages/bcc/__init__.py", line 485, in get_table
    keytype = BPF._decode_table_type(json.loads(key_desc))
  File "/usr/lib64/python3.6/site-packages/bcc/__init__.py", line 439, in _decode_table_type
    fields.append((t[0], BPF._decode_table_type(t[1])))
  File "/usr/lib64/python3.6/site-packages/bcc/__init__.py", line 454, in _decode_table_type
    raise Exception("Failed to decode type %s" % str(t))
Exception: Failed to decode type FOO
```

Support for enums in the generated JSON was added in commit f7f873a2 "Fix segfault with enumerations", but the added test doesn't actually reference the table from the Python side, so actually parsing the generated JSON in Python seems to have never been tested. Adding `t = b['drops']` to the test causes it to fail.

The `desc` for the enum is `['E', ['FOO'], 'enum']` and `BPF._decode_table_type()` iterates over `desc[1]`, assuming that each element is [a list] of length 2 or 3. When the element is a longer string instead, it fails.

I'm not sure if `libbcc.bpf_table_key_desc()` is [meant to be] used internally only, or if it might be used by users of bcc too, so I don't know if this should be fixed on the C++ or Python side. (Ie. can we just change the format of the generated JSON "no questions asked")

If we can just change the JSON, I'd propose doing that: Currently, the list of enumeration members is nearly useless anyway, since they don't include the associated values - I'd either remove the members completely, or add the associated values (though I expect that enums will just be represented as `ctypes.c_uint` or similar, so the members would be ignored in the Python code anyway).

There's also the question of size: Enum size is implementation-defined, so it depends on the compiler (some compilers automatically select the smallest type where all the values fit), the compiler flags (`-fshort-enums`), enum attributes (`packed`) and the range of values.

The simplest thing to do seems to be to change the generated JSON from `["fieldname", ["ENUM_MEMBER", ...], "enum"]` to `["fieldname", "unsigned int"]` (or similar), by replacing the code in `VisitEnumDecl()` with `TraverseType(D->getIntegerType()); return false;`. (That fixes the issue for me)

Speaking of `Traverse*()`, it seems that the root cause of the segfault that was referenced in the commit is calling `TraverseDecl()` from `VisitTagType()` (infinite recursion through `BMapDeclVisitor::VisitTagType -> RecursiveASTVisitor::TraverseDecl -> TraverseEnumDecl -> TraverseType -> TraverseEnumType -> WalkUpFromEnumType -> WalkUpFromTagType -> BMapDeclVisitor::VisitTagType`) - the [documentation in AST/RecursiveASTVisitor.h](https://github.com/llvm-mirror/clang/blob/ea521aa6024431138f7890d486a6e3500822b3d5/include/clang/AST/RecursiveASTVisitor.h#L111) seems to say that `Visit*()` may not call `Traverse*()` (nor `WalkUpFrom*()`), so since we're doing that in many places, it might be opening up a can of worms.

I'm not that familiar with clang, so I don't know what would be the best way to avoid the whole problem - maybe the intention is that you'd dispatch to a separate class when you want to call `Traverse*()` from `Visit*()`. You could also get the same effect by tracking a "traversal mode" and only doing things in `Visit*()` if you're in the right mode, though that'd quite likely just end up being messier, since it'd essentially be just a trick to "avoid new classes at any cost".

As for the enum size thing, both gcc and clang seem to use `unsigned int` by default (`int` if there's negative values), but since it's compiler dependent - and we almost certainly aren't using the same compiler that the kernel was compiled with (since we're using clang, whereas the kernel is usually built with gcc) - I think if the compiler used for the kernel were to use a different-sized enum (that wasn't due to `__attribute__((packed))`), we'd end up getting the size wrong. (Or does bcc extract the actual sizes from the kernel, and somehow feed them to clang?)

**In short:**
- There exists an easy fix (using `TraverseType()`)
- Calling `Traverse*()` from `Visit*()` seems to be bad
- In some (probably highly unlikely) configurations, we'll probably get enum sizes wrong

If you want the easy fix, I can make a pull request right away - otherwise, figuring out what is the proper fix will probably need more discussion (or for someone who can see the proper fix without discussion to fix it :P)

**Comment by yonghong-song** [comment]
@alexer thanks for great analysis. I think there is no need to expose the enumerators (values) in the JSON. They are not needed and actually complicating things. Replacing it with an appropriate int type is probably the right thing to do.

The string representation of key/value types can be used by both python and lua to reconstruct key/value type in python and lua, so the accuracy of type str is needed. But replacing enum type with proper int type (including signedness and type size) should be okay.

Could you help prepare a pull request?

**Comment by yonghong-song** [comment]
Regarding packed data structure. The offset should have already reflected it. 
https://clang.llvm.org/doxygen/classclang_1_1EnumDecl.html
I did not check. Maybe EnumDecl->getIntegerType() can give us enough information?

**Comment by alexer** [comment]
Yeah, I'll make the pull request.

> Regarding packed data structure. The offset should have already reflected it.

Oh, I didn't mean the offset (which changes when `packed` is applied to the containing struct), I meant the size of the enum (which changes when `packed` is applied to the enum itself - and in the other cases I listed).

`EnumDecl->getIntegerType()` does give the "right" size, in that it gives the size that clang used, but  that size isn't guaranteed the same that the kernel is using. (So this is really a completely separate issue from the JSON output - and a highly unlikely scenario in the first place)

If the enum size is modified by `__attribute__((packed))`, that's visible in the source and clang can see it, so we get it right - but if the kernel has been compiled with `-fshort-enums` (or with some compiler which does that by default - since it's implementation-defined), then when we ask clang about the size of the enum, it tells us what it thinks it *should* be, not what it actually is in the kernel, so we'd get the size wrong (and offsets for later fields, too, depending on alignment).

But like I said, this seems like a highly unlikely scenario, since in practice the kernel is always compiled with gcc, which has int-sized enums by default, like clang - and I doubt anyone would actually compile their kernel with `-fshort-enums`. (Though I'm not sure if non-gcc might be used in practice on some other architecture - also, it seems on *some* platforms, short enums are the default)

It might be that the enum size is in practice always dictated by the system ABI, so we're "protected in practice", but at least the [kernel folks seem to think it's a good idea to avoid enums in userspace APIs](https://lore.kernel.org/patchwork/patch/755514/), since the size might be different in different compilers - unfortunately, that doesn't help us, since we're dealing with internal kernel structures, which do include enums. :)

(FWIW, I *did* try compiling a kernel with `-fshort-enums` *"for science"*; It seems there's a few places (I found 4, 2 of them under arch/) that expect int-sized enums. When using bcc with that kernel, as expected it got the enum sizes wrong and returned garbage values for kernel structs with suitably placed enums, unless `bcc.BPF` was given `cflags=['-fshort-enums']`, too)

**Comment by yonghong-song** [comment]
@alexer Thanks for detailed explanation! I guess most distros if not all do not use `-fshort-enums` so we should be fine. Otherwise, indeed, we will have issues to match bpf program vs. kernel data structures unless `-fshort-enums` is also used to compile bcc.

I think for json visitor, using `EnumDecl->getIntegerType()` should be okay. If a particular kernel is compiling with `-fshort-enums`, bpf program should be added with additional `cflags` as you indicated in the above.
